File: frontend/src/types/factor.ts
export type FactorCategory =
  | 'momentum'
  | 'value'
  | 'quality'
  | 'growth'
  | 'volatility'
  | 'liquidity'
  | 'size'
  | 'technical'
  | 'custom';

export interface Factor {
  id: number;
  name: string;
  code: string;
  category: FactorCategory;
  description: string | null;
  formula: string | null;
  is_builtin: boolean;
  created_by: number | null;
  created_at: string;
  updated_at: string;
}

export interface FactorListResponse {
  items: Factor[];
  total: number;
  page: number;
  size: number;
}

export interface FactorListParams {
  page?: number;
  size?: number;
  category?: FactorCategory;
  keyword?: string;
}

export interface FactorCreate {
  name: string;
  code: string;
  category: FactorCategory;
  description?: string;
  formula?: string;
}

export interface FactorUpdate {
  name?: string;
  code?: string;
  category?: FactorCategory;
  description?: string;
  formula?: string;
}

export interface FactorCalculateRequest {
  factor_id: number;
  stock_codes?: string[];
  start_date: string;
  end_date: string;
}

export interface FactorCalculateResponse {
  factor_id: number;
  total_stocks: number;
  total_dates: number;
  calculated_count: number;
  status: string;
}

export interface FactorEvaluationRequest {
  factor_id: number;
  start_date: string;
  end_date: string;
  benchmark?: string;
}

export interface GroupReturnItem {
  group: number;
  return_value: number;
  stock_count: number;
}

export interface FactorEvaluation {
  id: number;
  factor_id: number;
  start_date: string;
  end_date: string;
  ic_mean: number | null;
  ic_std: number | null;
  ir: number | null;
  ic_positive_ratio: number | null;
  turnover: number | null;
  created_at: string;
}

export interface FactorEvaluationDetail {
  evaluation: FactorEvaluation;
  ic_series: Array<{ date: string; ic: number }>;
  group_returns: GroupReturnItem[];
}

export interface FactorCategoryStats {
  category: FactorCategory;
  count: number;
}

export const FACTOR_CATEGORY_LABELS: Record<FactorCategory, string> = {
  momentum: '动量',
  value: '估值',
  quality: '质量',
  growth: '成长',
  volatility: '波动',
  liquidity: '流动性',
  size: '规模',
  technical: '技术',
  custom: '自定义',
};

export interface FactorAnalyzeRequest {
  factor_id: number;
  start_date: string;
  end_date: string;
  stock_codes?: string[];
  force_recalculate?: boolean;
}

export interface FactorAnalyzeResponse {
  factor_id: number;
  calculated_count: number;
  evaluation: FactorEvaluation;
  ic_series: Array<{ date: string; ic: number }>;
  group_returns: GroupReturnItem[];
}

File: backend/app/schemas/factor.py
from datetime import date, datetime
from enum import Enum
from typing import List, Optional

from pydantic import BaseModel, ConfigDict, Field


class FactorCategory(str, Enum):
    MOMENTUM = "momentum"
    VALUE = "value"
    QUALITY = "quality"
    GROWTH = "growth"
    VOLATILITY = "volatility"
    LIQUIDITY = "liquidity"
    SIZE = "size"
    TECHNICAL = "technical"
    CUSTOM = "custom"


class FactorBase(BaseModel):
    name: str = Field(..., min_length=1, max_length=100)
    code: str = Field(..., min_length=1, max_length=50)
    category: FactorCategory = FactorCategory.CUSTOM
    description: Optional[str] = None
    formula: Optional[str] = None


class FactorCreate(FactorBase):
    pass


class FactorUpdate(BaseModel):
    name: Optional[str] = Field(None, min_length=1, max_length=100)
    code: Optional[str] = Field(None, min_length=1, max_length=50)
    category: Optional[FactorCategory] = None
    description: Optional[str] = None
    formula: Optional[str] = None


class FactorResponse(FactorBase):
    model_config = ConfigDict(from_attributes=True)

    id: int
    is_builtin: bool
    created_by: Optional[int] = None
    created_at: datetime
    updated_at: datetime


class FactorListResponse(BaseModel):
    items: List[FactorResponse]
    total: int
    page: int
    size: int


class FactorValueCreate(BaseModel):
    stock_code: str
    date: date
    value: Optional[float] = None


class FactorValueResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    factor_id: int
    stock_code: str
    date: datetime
    value: Optional[float] = None


class FactorValueListResponse(BaseModel):
    items: List[FactorValueResponse]
    total: int


class FactorCalculateRequest(BaseModel):
    factor_id: int
    stock_codes: Optional[List[str]] = None
    start_date: date
    end_date: date


class FactorCalculateResponse(BaseModel):
    factor_id: int
    total_stocks: int
    total_dates: int
    calculated_count: int
    status: str


class FactorEvaluationResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    factor_id: int
    start_date: datetime
    end_date: datetime
    ic_mean: Optional[float] = None
    ic_std: Optional[float] = None
    ir: Optional[float] = None
    ic_positive_ratio: Optional[float] = None
    turnover: Optional[float] = None
    created_at: datetime


class FactorEvaluationRequest(BaseModel):
    factor_id: int
    start_date: date
    end_date: date
    benchmark: Optional[str] = "000300"


class GroupReturnItem(BaseModel):
    group: int
    return_value: float
    stock_count: int


class FactorEvaluationDetailResponse(BaseModel):
    evaluation: FactorEvaluationResponse
    ic_series: List[dict]
    group_returns: List[GroupReturnItem]


class FactorCategoryStats(BaseModel):
    category: FactorCategory
    count: int


class FactorAnalyzeRequest(BaseModel):
    factor_id: int
    start_date: date
    end_date: date
    stock_codes: Optional[List[str]] = None
    force_recalculate: bool = False


class FactorAnalyzeResponse(BaseModel):
    factor_id: int
    calculated_count: int
    evaluation: FactorEvaluationResponse
    ic_series: List[dict]
    group_returns: List[GroupReturnItem]
